accounts/abi/bind/topics.go

```diff
@@ -103,151 +103,61 @@ func makeTopics(query ...[]interface{}) ([][]common.Hash, error) {
-				default:
-					return fmt.Errorf("unsupported indexed type: %v", arg.Type)
-				}
-			}
-		}
-		topics = topics[1:]
+		field.Set(reflect.ValueOf(reconstr))
 	}
-	return nil
+	return parseTopicIntoFunc(fields, topics, store)
 }
 
 // parseTopicsIntoMap converts the indexed topic field-value pairs into map key-value pairs
 func parseTopicsIntoMap(out map[string]interface{}, fields abi.Arguments, topics []common.Hash) error {
+	store := func(arg abi.Argument, reconstr interface{}) {
+		out[arg.Name] = reconstr
+	}
+	return parseTopicIntoFunc(fields, topics, store)
+}
+
+// parseTopicIntoFunc converts the indexed topic field-value pairs and stores them using the
+// provided store function.
+//
+// Note, dynamic types cannot be reconstructed since they get mapped to Keccak256
+// hashes as the topic value!
+func parseTopicIntoFunc(fields abi.Arguments, topics []common.Hash, store func(abi.Argument, interface{})) error {
 	// Sanity check that the fields and topics match up
 	if len(fields) != len(topics) {
 		return errors.New("topic/field count mismatch")
 	}
 	// Iterate over all the fields and reconstruct them from topics
-	for _, arg := range fields {
+	for i, arg := range fields {
 		if !arg.Indexed {
 			return errors.New("non-indexed field in topic reconstruction")
 		}
-
+		//reconstr, err := parseField(arg, topics[i])
+		var reconstr interface{}
 		switch arg.Type.T {
-		case abi.BoolTy:
-			out[arg.Name] = topics[0][common.HashLength-1] == 1
-		case abi.IntTy, abi.UintTy:
-			out[arg.Name] = abi.ReadInteger(arg.Type.T, arg.Type.Kind, topics[0].Bytes())
-		case abi.AddressTy:
-			var addr common.Address
-			copy(addr[:], topics[0][common.HashLength-common.AddressLength:])
-			out[arg.Name] = addr
-		case abi.HashTy:
-			out[arg.Name] = topics[0]
-		case abi.FixedBytesTy:
-			array, err := abi.ReadFixedBytes(arg.Type, topics[0].Bytes())
-			if err != nil {
-				return err
-			}
-			out[arg.Name] = array
 		case abi.StringTy, abi.BytesTy, abi.SliceTy, abi.ArrayTy:
 			// Array types (including strings and bytes) have their keccak256 hashes stored in the topic- not a hash
 			// whose bytes can be decoded to the actual value- so the best we can do is retrieve that hash
-			out[arg.Name] = topics[0]
+			reconstr = topics[i]
 		case abi.FunctionTy:
-			if garbage := binary.BigEndian.Uint64(topics[0][0:8]); garbage != 0 {
-				return fmt.Errorf("bind: got improperly encoded function type, got %v", topics[0].Bytes())
+			if garbage := binary.BigEndian.Uint64(topics[i][0:8]); garbage != 0 {
+				return fmt.Errorf("bind: got improperly encoded function type, got %v", topics[i].Bytes())
 			}
 			var tmp [24]byte
-			copy(tmp[:], topics[0][8:32])
-			out[arg.Name] = tmp
-		default: // Not handling tuples
-			return fmt.Errorf("unsupported indexed type: %v", arg.Type)
+			copy(tmp[:], topics[i][8:32])
+			reconstr = tmp
+		default:
+			var err error
+			reconstr, err = abi.ToGoType(0, arg.Type, topics[i].Bytes())
```

Comment:
Since you're changing the behavior here by supporting tupples, you should add a unit test to check that it works.